stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -172,64 +212,245 @@ public void run() {
     });
   }
 
-  private Node createNodeForElement(Object element) {
-    mDOMProvider.verifyThreadAccess();
-
-    NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(element);
-
-    Node node = new Node();
-    node.nodeId = mObjectIdMapper.putObject(element);
-    node.nodeType = descriptor.getNodeType(element);
-    node.nodeName = descriptor.getNodeName(element);
-    node.localName = descriptor.getLocalName(element);
-    node.nodeValue = descriptor.getNodeValue(element);
+  private void updateTree() {
+    //
+    long startTimeMs = SystemClock.elapsedRealtime();
+    //
 
-    node.children = getChildNodesForElement(element);
-    node.childNodeCount = node.children.size();
+    mDOMProvider.verifyThreadAccess();
 
-    node.attributes = new ArrayList<String>();
-    descriptor.copyAttributes(element, new AttributeListAccumulator(node.attributes));
+    ShadowDOM.Update domUpdate = createShadowDOMUpdate();
+    boolean isEmpty = domUpdate.isEmpty();
+    if (isEmpty) {
+      domUpdate.abandon();
+    } else {
+      applyDOMUpdate(domUpdate);
+    }
 
-    return node;
+    //
+    long endTimeMs = SystemClock.elapsedRealtime();
+    long deltaMs = endTimeMs - startTimeMs;
+    LogUtil.e(String.format(
+        "DOM.updateTree() completed in %s ms%s",
+        Long.toString(deltaMs),
+        isEmpty ? " (no changes)" : ""));
+    //
   }
 
-  private List<Node> getChildNodesForElement(Object element) {
+  private ShadowDOM.Update createShadowDOMUpdate() {
     mDOMProvider.verifyThreadAccess();
 
-    NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(element);
-    int childNodeCount = descriptor.getChildCount(element);
+    if (mDOMProvider.getRootElement() != mShadowDOM.getRootElement()) {
+      throw new IllegalStateException();
+    }
 
-    List<Node> childNodes;
-    if (childNodeCount == 0) {
-      childNodes = Collections.emptyList();
-    } else {
-      childNodes = new ArrayList<Node>(childNodeCount);
-      for (int i = 0; i < childNodeCount; ++i) {
-        Object childElement = descriptor.getChildAt(element, i);
-        Node childNode = createNodeForElement(childElement);
-        childNodes.add(childNode);
+    ShadowDOM.UpdateBuilder updateBuilder = mShadowDOM.beginUpdate();
+    ListAccumulator<Object> childrenAccumulator = acquireChildrenAccumulator();
+    Queue<Object> queue = new ArrayDeque<>();
+    queue.add(mDOMProvider.getRootElement());
+
+    while (!queue.isEmpty()) {
+      final Object element = queue.remove();
+      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(element);
+      mObjectIdMapper.putObject(element);
+      descriptor.getChildren(element, childrenAccumulator);
+      updateBuilder.setElementChildren(element, childrenAccumulator);
+      for (int i = 0, N = childrenAccumulator.size(); i < N; ++i) {
+        queue.add(childrenAccumulator.get(i));
       }
+      childrenAccumulator.clear();
```

Comment:
Seems weird that you don't just create a new accumulator that delegates to the queue then pass that in?